blog/content/post/2024-01-31-drm-bridge-chaining.md

```diff
@@ -0,0 +1,863 @@
+
+```shell
+# cat /sys/bus/i2c/devices/*c/name
+sn65dsi84
+(...)
+it6263
+# dmesg | grep it6263
+[    3.130625] it6263_probe has been called
+[    3.131740] it6263_reset has been called
+(...)
+[    3.694544] it6263_bridge_attach has been called
+(...)
+# dmesg | grep sn65dsi84
+[    3.658664] sn65dsi83_probe has been called
+[    3.661588] sn65dsi83_parse_dt has been called
+(...)
+[    3.689746] sn65dsi83_attach has been called
+(...)
+```
+
+> Note: `dmesg` logs from above are provided by adding `printk()` to the drivers
+source code.
+
+After further analysis, the following problems were pointed out:
+
+* No video output on the panel during boot as well as in userspace;
+* Weston argues that one `drm_connector` does not have CRTC assigned:
+
+    ```shell
+    [11:44:08.403] weston 10.0.1
+    (...)
+    [11:44:09.436] Output 'DSI-1' using color profile: built-in default sRGB SDR profile
+    [11:44:09.444] Output DSI-1 (crtc 33) video modes:
+                   1280x720@59.9 16:9, current, 74.2 MHz
+    [11:44:09.444] Output 'DSI-1' enabled with head(s) DSI-1
+    [11:44:09.444] Module '/usr/lib/libgbm.so' already loaded
+    [11:44:09.444] Output 'HDMI-A-1' using color profile: built-in default sRGB SDR profile
+    [11:44:09.444] Output 'HDMI-A-1': No available CRTCs.
+    [11:44:09.444] Enabling output "HDMI-A-1" failed.
+    [11:44:09.444] Error: cannot enable output 'HDMI-A-1' without heads.
+    ```
+
+[sn65dsi84-varigit-repo]: https://github.com/varigit/linux-imx/tree/lf-5.10.y_var04/drivers/gpu/drm/bridge/sn65dsi83
+[sn65dsi84-nxp-repo]: https://github.com/nxp-imx/linux-imx/blob/lf-5.15.y/drivers/gpu/drm/bridge/ti-sn65dsi83.c
+[drm_bridge_attach-bootlin]: https://elixir.bootlin.com/linux/v5.15.71/source/drivers/gpu/drm/drm_bridge.c#L175
+
+##### Weston issue
+
+According to [Linux Kernel DRM/KMS system documentation][kms-linux-docs] the KMS
+structure should be following:
+
+![kms-structure](/img/kms-structure-bridge-chaining.png)
+
+The red-marked `drm_connector` structure is surplus and is causing the Weston
+issues.
+
+According to [comments in `drm_bridge_connector.c`][drm-bridge-connector-com]
+and to [comments in `drm_bridge.c`][drm-bridge-com] file, to create
+`drm_connector` for a chain of bridges, DRM drivers should match the following
+requirements:
+
+* Encoder or display controller drivers should:
+  * Create `drm_connector` structure and initialize it by using
+  `drm_bridge_connector_init()` function from `drm_bridge_connector.c`;
+* Bridge drivers should:
+  * Export following functions via `drm_bridge_funcs` (some of them are
+  optional): `drm_bridge_funcs.detect()`, `drm_bridge_funcs.get_modes()`,
+  `drm_bridge_funcs.get_edid()`, `drm_bridge_funcs.get_edid()`,
+  `drm_bridge_funcs.hpd_enable()` and `drm_bridge_funcs.hpd_disable().`;
+  * Export its type and functionalities via flags `drm_bridge.type`and
+  `drm_bridge.ops`
+  * Do not create connectors by themselves or implement conditional
+  connector creation.
```

Comment:
Bullet points should be read as a continuation of higher-level bullet point sentences to keep consistency. If I read `Bridge drivers should: ... Do not ...` it doesn't sound consistent.

Reply:
Fixed here 92b439ca1c263d38890f86404b8af0bd8834c42d .